pkg/naming/metrics_query_test.go

```diff
@@ -340,6 +342,36 @@ func TestBuildExternalSelector(t *testing.T) {
 				hasSelector(`foo="bar",namespaces="default"`),
 			),
 		},
+		{
+			name: "single LabelMatchers value with namespace - detached namespace",
+
+			mq:        mustNewQuery(`<<.LabelMatchers>>`),
+			namespace: "default",
+			options:   config.Options{DetatchFromNamespace: true},
+			metricSelector: labels.NewSelector().Add(
+				*mustNewLabelRequirement("foo", selection.Equals, []string{"bar"}),
+			),
+
+			check: checks(
+				hasError(nil),
+				hasSelector(`foo="bar"`),
+			),
+		},
+		{
+			name: "single LabelMatchers value with namespace - explicit attached namespace",
+
+			mq:        mustNewQuery(`<<.LabelMatchers>>`),
+			namespace: "default",
+			options:   config.Options{DetatchFromNamespace: false},
+			metricSelector: labels.NewSelector().Add(
+				*mustNewLabelRequirement("foo", selection.Equals, []string{"bar"}),
```

Comment:
can you add a test where a different namespace label is provided?

Reply:
Good call. Have added additional test case to ensure there is no 'default' namespace applied when `config.Options{DetatchFromNamespace} == false`.

Reply:
Can you add another one where the user specifies a label requirement for the namespace label, and we ensure that it's truly enforced?

Reply:
Hi @brancz ,
Apologies for the delay..

Had a think about your request, my understanding is that you would like to ensure that when a user specifies a Label requirement that has key `namespace`, that this should also be dropped while forming the selector.

In this case, I would suggest not enforcing this.

**Reasoning**:
The original intention of this feature is to allow cross namespace metrics, ensuring that metrics are not restricted to the namespace to which the HPA resource  resides.
If we had a scenario where we are doing cross-namespace metrics, and the metric we are retrieving happens to have a `namepace` label that we need to target (for whatever reason) then, enforcing detached namespace in this scenario would prevent the user from fulfilling their requirements.

What are your thoughts?

Reply:
@AshMenhennett I am very sorry for the long delay, but I am still having doubts here. While I agree that external metrics are not bound to namespaces, having an explicit option to allow inter-namespace cross-correlation still sounds like a leaky abstraction.

Wouldn't be your problem be solvable by simply adding a Prometheus recording rule to your stack which cross-correlates namespaces and reference that in the adapter configuration?

Reply:
The external metrics API was intentionally specified to be bound to a namespace, in this case it doesn't map 100% perfectly against Prometheus norms, but I think violating this would be a mistake. If we can handle this with a recording rule instead I do agree that I would prefer that approach.

Reply:
Thanks for the timely reply guys!

I definitely see your concerns and understand the intention of the specification and the need to stay inline with it.

As a tactical solution, Prometheus recording rules could be used, however, as a more strategic solution, the ability to allow fine-grained control on a per resource case would be the primary goal/ benefit.

For example, in a large enough setting, this would require recording rules and the queries to be maintained for each scaling resource, each metric source and each environment. This isn't a massive issue, but I'm sure you can see what I'm getting at.

Thoughts?

Reply:
Are we at consensus of leveraging recording rules rather than introducing this feature?

Reply:
@AshMenhennett: yes, that is our current suggested solution. Please let me know if this works out for you. I am really in favor of a recording rule based solution as we would not have to introduce additional code paths.

Another data point is that configuration of prometheus-adapter is already complex enough, hence I am voting to add code if and only if we have no workaround available in Prometheus configuration.